src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -513,26 +510,159 @@ public void ClearValidationState(string key)
+        private static int GetArrayIndexDelimiter(string key, int index)
+        {
+            if (key[index] == '[')
+            {
+                var closeParanthesesIndex = key.IndexOf(']', index + 1);
+                if (closeParanthesesIndex == -1)
+                {
+                    // Key: "person.Addresses[0.BadKey" Consume the key in entirety.
+                    return -1;
+                }
+                else if (closeParanthesesIndex < key.Length - 1)
+                {
+                    if (key[closeParanthesesIndex + 1] == '.')
+                    {
+                        // person.Addresses[0].Street
+                        // Create a prefix node for 'Addresses[0].'
+                        index = closeParanthesesIndex + 1;
+                    }
+                    else
+                    {
+                        // person.Addresses[0]Street. Consume the key in entirety
+                        return -1;
+                    }
+                }
+                else
+                {
+                    // person.CarIds[17]. Consume the key in entirety.
+                    return -1;
+                }
+            }
+
+            return index;
+        }
+
+        private ModelStateNode FindNode(string key)
+        {
+            if (key.Length == 0)
+            {
+                return _root;
+            }
+
+            var currentRoot = _root;
+            var previousIndex = 0;
+            int index;
+            while ((index = key.IndexOfAny(Delimiters, previousIndex)) != -1)
+            {
+                index = GetArrayIndexDelimiter(key, index);
+                if (index == -1)
+                {
+                    break;
+                }
+
+                var subKey = new StringSegment(key, previousIndex, index - previousIndex);
+                var next = GetChildNode(currentRoot, subKey);
+                if (next == null)
+                {
+                    return null;
+                }
+
+                currentRoot = next;
+                previousIndex = index + 1;
+            }
+
+            if (previousIndex < key.Length - 1)
+            {
+                var subKey = new StringSegment(key, previousIndex, key.Length - previousIndex);
+                currentRoot = GetChildNode(currentRoot, subKey);
+            }
+
+            return currentRoot;
+        }
+
+        private static ModelStateNode GetChildNode(ModelStateNode currentRoot, StringSegment subKey)
         {
+            if (currentRoot.Children == null)
+            {
+                return null;
+            }
+
+            var searchKey = new KeyValuePair<StringSegment, ModelStateNode>(subKey, value: null);
+            var index = currentRoot.Children.BinarySearch(searchKey, ModelStateNodeComparer.Instance);
+            return index < 0 ? null : currentRoot.Children[index].Value;
+        }
 
+        private static ModelStateNode GetOrAddChildNode(
+            ModelStateNode root,
```

Comment:
This isn't the root, it's the parent - you can make this a method on `ModelStateNode`